api/v1alpha4/clusterclass_webhooks.go

```diff
@@ -0,0 +1,154 @@
+		in.Spec.Infrastructure.Ref.Namespace = in.Namespace
+	}
+	if len(in.Spec.ControlPlane.Ref.Namespace) == 0 {
+		in.Spec.ControlPlane.Ref.Namespace = in.Namespace
+	}
+	for i := range in.Spec.Worker.Deployments {
+		if len(in.Spec.Worker.Deployments[i].Template.Bootstrap.Ref.Namespace) == 0 {
+			in.Spec.Worker.Deployments[i].Template.Bootstrap.Ref.Namespace = in.Namespace
+		}
+		if len(in.Spec.Worker.Deployments[i].Template.Infrastructure.Ref.Namespace) == 0 {
+			in.Spec.Worker.Deployments[i].Template.Infrastructure.Ref.Namespace = in.Namespace
+		}
+	}
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateCreate() error {
+	return in.validate(nil)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateUpdate(old runtime.Object) error {
+	oldClusterClass, ok := old.(*ClusterClass)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a ClusterClass but got a %T", old))
+	}
+	return in.validate(oldClusterClass)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateDelete() error {
+	return nil
+}
+
+func (in *ClusterClass) validate(old *ClusterClass) error {
+	var allErrs field.ErrorList
+
+	// ensure all the references are within the same namespace
+	if in.Spec.Infrastructure.Ref.Namespace != in.Namespace {
+		allErrs = append(
+			allErrs, field.Invalid(field.NewPath("spec", "infrastructure", "ref", "namespace"), in.Spec.Infrastructure.Ref.Namespace, "must match metadata.namespace"),
+		)
+	}
+	if in.Spec.ControlPlane.Ref.Namespace != in.Namespace {
+		allErrs = append(
+			allErrs, field.Invalid(field.NewPath("spec", "controlPlane", "ref", "namespace"), in.Spec.ControlPlane.Ref.Namespace, "must match metadata.namespace"),
+		)
+	}
+	for _, class := range in.Spec.Worker.Deployments {
+		if class.Template.Bootstrap.Ref.Namespace != in.Namespace {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("ClusterClass", "spec", "worker", "deployments", "template", "bootstrap", "ref", "namespace"), class.Template.Bootstrap.Ref.Namespace, "must match metadata.namespace"),
+			)
+		}
+		if class.Template.Infrastructure.Ref.Namespace != in.Namespace {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("ClusterClass", "spec", "worker", "deployments", "infrastructure", "bootstrap", "ref", "namespace"), class.Template.Infrastructure.Ref.Namespace, "must match metadata.namespace"),
+			)
+		}
+	}
+
+	// Ensure MachineDeployment class are unique.
+	classNames := make(map[string]struct{})
+	for _, class := range in.Spec.Worker.Deployments {
+		if _, ok := classNames[class.Class]; ok {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("ClusterClass", "spec", "worker", "deployments"), class, fmt.Sprintf("MachineDeployment class %q is defined more than once.", class.Class)),
```

Comment:
I see where the earlier message in the cluster_webhook came from :)

Reply:
All fixed! 😉